### PR TITLE
[serve] Add replica states to status

### PR DESCRIPTION
**Pull request by zcin**
## Why are these changes needed?

Add number of running replicas to `serve status` and `serve.status()`.
Example output `serve status`:
```
proxies:
  090fe807117c1d14aa3eedfe734467aaa95915f893a80a4b01016588: HEALTHY
applications:
  default:
    status: RUNNING
    message: ''
    last_deployed_time_s: 1691541115.2581592
    deployments:
      default_Hello:
        status: HEALTHY
        replica_states:
          RUNNING: 3
        message: ''
```

Deployment fail:
```
proxies:
  a82df9b78f4a8db4de9d4af5ea455d4be4f3ec8bed95fe2acb130c79: HEALTHY
applications:
  default:
    status: DEPLOY_FAILED
    message: The deployments ['default_Hello'] are UNHEALTHY.
    last_deployed_time_s: 1691541166.991966
    deployments:
      default_Hello:
        status: UNHEALTHY
        replica_states: {}
        message: |-
          The deployment failed to start 10 times in a row. This may be due to a problem with its constructor or initial health check failing. See controller logs for details. Retrying after 2.0 seconds. Error:
          ray::ServeReplica:default_Hello.initialize_and_get_metadata() (pid=54147, ip=10.104.98.7, actor_id=53f03e7cf2fc24286b48755401000000, repr=<ray.serve._private.replica.ServeReplica:default_Hello object at 0x10b0c7a30>)
            File "/Users/cindyz/miniforge3/envs/ray/lib/python3.9/concurrent/futures/_base.py", line 439, in result
              return self.__get_result()
            File "/Users/cindyz/miniforge3/envs/ray/lib/python3.9/concurrent/futures/_base.py", line 391, in __get_result
              raise self._exception
            File "/Users/cindyz/ray/python/ray/serve/_private/replica.py", line 419, in initialize_and_get_metadata
              raise RuntimeError(traceback.format_exc()) from None
          RuntimeError: Traceback (most recent call last):
            File "/Users/cindyz/ray/python/ray/serve/_private/replica.py", line 407, in initialize_and_get_metadata
              await self._initialize_replica()
            File "/Users/cindyz/ray/python/ray/serve/_private/replica.py", line 184, in initialize_replica
              await sync_to_async(_callable.__init__)(*init_args, **init_kwargs)
            File "/Users/cindyz/workspace/hello.py", line 18, in __init__
              1/0
          ZeroDivisionError: division by zero
```

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/schema.py [hunk 1]
```diff
@@ -621,6 +621,7 @@ def get_empty_schema_dict() -> Dict:
 @dataclass
 class DeploymentStatusOverview:
     status: DeploymentStatus
+    num_running_replicas: int
```
Comment (zcin) [inline python/ray/serve/schema.py:624]: I named this `num_running_replicas` so it's not confused with `num_replicas` set in the config, but no strong opinion if folks think `num_replicas` is better!

**Review by shrekris-anyscale — COMMENTED** [review]
Does `num_running_replicas` include only replicas from the latest deployment version?

**Comment by zcin** [comment]
> Does `num_running_replicas` include only replicas from the latest deployment version?

No, it includes all replicas that are in the `RUNNING` state.

**Comment by edoakes** [comment]
@zcin instead of `num_running_replicas,` should we make it a dictionary of state: count? Shouldn't be any harder to add and is more useful/general

**Comment by zcin** [comment]
> @zcin instead of `num_running_replicas,` should we make it a dictionary of state: count? Shouldn't be any harder to add and is more useful/general

That's a good idea, I've renamed the new field to `replicas` and its type is a dictionary of replica state -> count.

**Comment by edoakes** [comment]
@zcin sorry for the back and forth, but let's name it something slightly more specific such as `replica_states`. The reasoning is that I can see us wanting to add more detail about the replicas in the future under a `replicas` field, so if we use that name now it will cause us problems if/when we do that.

**Review by edoakes — APPROVED** [review]

**Pull request merged**